Select shared memory parallelization

philbucher opened the pull request:
**Description**
This PR adds a mechanism for selecting the shared memory parallization (SMP)
I took a close look at how [blaze](https://bitbucket.org/blaze-lib/blaze/src/af002e5818cd28a15fdea3376c203024b1e55be9/CMakeLists.txt#lines-294:330) handles this and adapted it for our case (I felt we can use a simpler solution with only one flag)

now one can specify `KRATOS_SHARED_MEMORY_PARALLELIZATION` to `OpenMP`, `C++11` or `None` (can be extended if necessary)

In the code we can use `KRATOS_SMP_OPENMP`, `KRATOS_SMP_CXX11` and `KRATOS_SMP_NONE` 

This would mostly be used in the Parallel Utilities

**Changelog**
Adding selection mechanism for SMP

philbucher committed:
adding mechanism

philbucher committed:
migrate openmp

philbucher committed:
error in case old option is used

philbucher committed:
cleanup

roigcarlo commented on CMakeLists.txt:
```diff
@@ -284,25 +284,42 @@ endif(CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT)
+endif(${KRATOS_EXCLUDE_OPENMP} MATCHES ON)
+
+# default to OpenMP
+set(KRATOS_SHARED_MEMORY_PARALLELIZATION "OpenMP" CACHE STRING "Specify which thread library is used for shared-memory parallelization.")
```
Maybe change this to something like:

```CMake
# Link on install
if(NOT DEFINED KRATOS_SHARED_MEMORY_PARALLELIZATION)
  SET(KRATOS_SHARED_MEMORY_PARALLELIZATION "OpenMP")
endif(NOT DEFINED KRATOS_SHARED_MEMORY_PARALLELIZATION)
```

Not really fan of the cache mechanism to handle undefined variables, we had problems in the past with that and nobody really uses CMake GUI for Kratos anyway

philbucher replied:
done

philbucher committed:
remove warning disable

this should not be there, only hides potential problems
can still be done locally if needed

philbucher committed:
improve default
